src/main/java/jenkins/plugins/git/GitPublisherStep.java

```diff
@@ -0,0 +1,393 @@
+
+          if (remote == null)
+            throw new AbortException("No repository found for target repo name " + targetRepo);
+
+          // expand environment variables in remote repository
+          remote = gitSCM.getParamExpandedRepo(environment, remote);
+
+          listener.getLogger().println("Pushing HEAD to branch " + branchName + " at repo "
+            + targetRepo);
+          remoteURI = remote.getURIs().get(0);
+          PushCommand push = git.push().to(remoteURI).ref("HEAD:" + branchName);
+          if (forcePush) {
+            push.force();
+          }
+          push.execute();
+        } catch (GitException e) {
+          e.printStackTrace(listener.error("Failed to push branch " + branchName + " to " + targetRepo));
+        }
+      }
+    }
+
+    if (isPushNotes()) {
+      for (final GitPublisher.NoteToPush b : notesToPush) {
+        if (b.getnoteMsg() == null)
+          throw new AbortException("No note to push defined");
+
+        b.setEmptyTargetRepoToOrigin();
+        String noteMsgTmp = environment.expand(b.getnoteMsg());
+        final String noteMsg = replaceAdditionalEnvironmentalVariables(noteMsgTmp, run);
+        final String noteNamespace = environment.expand(b.getnoteNamespace());
+        final String targetRepo = environment.expand(b.getTargetRepoName());
+        final boolean noteReplace = b.getnoteReplace();
+
+        try {
+          // Lookup repository with unexpanded name as GitSCM stores them unexpanded
+          RemoteConfig remote = gitSCM.getRepositoryByName(b.getTargetRepoName());
+
+          if (remote == null) {
+            listener.getLogger().println("No repository found for target repo name " + targetRepo);
+          }
+
+          // expand environment variables in remote repository
+          remote = gitSCM.getParamExpandedRepo(environment, remote);
+
+          listener.getLogger().println("Adding note to namespace \""+noteNamespace +"\":\n" + noteMsg + "\n******" );
+
+          if ( noteReplace )
+            git.addNote(    noteMsg, noteNamespace );
+          else
+            git.appendNote( noteMsg, noteNamespace );
+
+          remoteURI = remote.getURIs().get(0);
+          PushCommand push = git.push().to(remoteURI).ref("refs/notes/*");
+          if (forcePush) {
+            push.force();
+          }
+          push.execute();
+        } catch (GitException e) {
+          e.printStackTrace(listener.error("Failed to add note: \n" + noteMsg  + "\n******"));
+        }
+      }
+    }
+  }
+
+  private String replaceAdditionalEnvironmentalVariables(String input, Run<?, ?> run){
+    if (run == null){
+      return input;
+    }
+    String buildResult = "";
+    Result result = run.getResult();
+    if (result != null) {
+      buildResult = result.toString();
+    }
+    String buildDuration = run.getDurationString().replaceAll("and counting", "");
+
+    input = input.replaceAll("\\$BUILDRESULT", buildResult);
```

Comment:
Should use the utility methods in core to do variable replacement (unless I'm missing something).

Also seems like a limited use case, what is this for exactly? This is a Pipeline step, why need environment variables at all?